### PR TITLE
Tbingol/delete slipped puts

### PR DESCRIPTION
**Pull request by tbingol**
Fix Issue #913: Delete blobs that were successfully uploaded to some replicas but failed to satisfy the quorum

PR maintains a set of chunk blob ids for the failed attempts during a put operation and issues delete requests on them upon the conclusion the operation.

### CODE REVIEW

File: ambry-router/src/main/java/com/github/ambry/router/BackgroundDeleteRequest.java [hunk 1]
```diff
@@ -24,6 +25,7 @@
   static final String SERVICE_ID_PREFIX = "ambry-background-delete-";
   private final StoreKey storeKey;
   private final String serviceId;
+  private final BlobId blobId;
```
Comment (ankagrawal) [inline ambry-router/src/main/java/com/github/ambry/router/BackgroundDeleteRequest.java:28]: BlobId is a StoreKey (it implements the storeKey interface). So why introduce both and set one or the other? Can't storeKey be used in every case?

File: ambry-router/src/main/java/com/github/ambry/router/RouterCallback.java [hunk 2]
```diff
@@ -60,5 +62,17 @@ void scheduleDeletes(List<StoreKey> idsToDelete, String serviceIdSuffix) {
       backgroundDeleteRequests.add(new BackgroundDeleteRequest(storeKey, serviceIdSuffix));
     }
   }
+
+  /**
+   * Schedule the deletes of ids in the given list.
+   * @param idsToDelete the set of blob ids that need to be deleted.
+   * @param serviceIdSuffix the suffix to append to the service ID when deleting these blobs.
+   */
+  void scheduleDeletesForBlobIds(Set<BlobId> idsToDelete, String serviceIdSuffix) {
```
Comment (ankagrawal) [inline ambry-router/src/main/java/com/github/ambry/router/RouterCallback.java:71]: maybe we can rename this to scheduleSlippedDeletes()? In fact we might not even need this method if we didn't change the `BackgroundDeleteRequest` class (see comments above).

**Review by ankagrawal — APPROVED** [review]
Look good to me overall. Can you please fix the tests. I guess just re-running them might help fix it -- failure don't seem to be related to this PR.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java [hunk 3]
```diff
@@ -715,6 +715,87 @@ public void onBlobDeleted(String blobId, String serviceId, Account account, Cont
     Assert.assertEquals("All operations should have completed", 0, router.getOperationsCount());
   }
 
+  /**
+   * Test that even when a composite blob put succeeds, the slipped put data chunks are deleted.
+   */
+  @Test
+  public void testSuccessfulPutDataChunkDelete() throws Exception {
```
Comment (cgtz) [inline ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:722]: We should monitor how this test case performs on CI runs. If it fails intermittently, we may have to make some changes.

I understand why it is probabilistic though.

**Review by cgtz — APPROVED** [review]

**Pull request merged**